crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,337 @@
+///
+/// Archetypes are only modified when a novel archetype (set of components) is seen for the first time;
+/// swapping between existing archetypes will not trigger these checks.
+///
+/// Note that this is converted to an [`UntypedArchetypeInvariant`] when added to a [`World`].
+/// This is to ensure compatibility between different invariants.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ArchetypeInvariant<B1: Bundle, B2: Bundle = B1> {
+    /// For all entities where the predicate is true
+    pub predicate: ArchetypeStatement<B1>,
+    /// The consequence must also be true
+    pub consequence: ArchetypeStatement<B2>,
+}
+
+impl<B1: Bundle, B2: Bundle> ArchetypeInvariant<B1, B2> {
+    /// Erases the type information of this archetype invariant.
+    ///
+    /// Requires mutable world access, since the components might not have been added to the world yet.
+    #[inline]
+    pub fn into_untyped(self, world: &mut World) -> UntypedArchetypeInvariant {
+        UntypedArchetypeInvariant {
+            predicate: self.predicate.into_untyped(world),
+            consequence: self.consequence.into_untyped(world),
+        }
+    }
+}
+
+impl<B: Bundle> ArchetypeInvariant<B, B> {
+    /// This is a helper function for constructing common invariants.
+    /// All components of the provided bundle require each other.
+    /// In other words, if any one component of this bundle is present, then all of them must be.
+    #[inline]
+    pub fn full_bundle() -> Self {
+        Self {
+            predicate: ArchetypeStatement::<B>::at_least_one_of(),
+            consequence: ArchetypeStatement::<B>::all_of(),
+        }
+    }
+}
+
+/// A statement about the presence or absence of some subset of components in the given [`Bundle`]
+///
+/// This type is used as part of an [`ArchetypeInvariant`].
+///
+/// When used as a predicate, the archetype invariant matches all entities which satisfy the statement.
+/// When used as a consquence, then the statment must be true for all entities that were matched by the predicate.
+///
+/// For the statements about a single component `C`, wrap it in a single-component bundle `(C,)`.
+/// For single component bundles, `AllOf` and `AtLeastOneOf` are equivalent.
+/// Prefer `ArchetypeStatement::<(C,)>::all_of` over `ArchetypeStatement::<(C,)>::at_least_one_of` for consistency and clarity.
+///
+/// Note that this is converted to an [`UntypedArchetypeStatement`] when added to a [`World`].
+/// This is to ensure compatibility between different invariants.
+#[derive(Clone, Debug, PartialEq)]
+pub enum ArchetypeStatement<B: Bundle> {
+    /// Evaluates to true if and only if the entity has all of the components present in the bundle `B`
+    AllOf(PhantomData<B>),
+    /// The entity has at least one component in the bundle `B`, and may have all of them.
+    /// When using a single-component bundle, `AllOf` is preferred.
+    AtLeastOneOf(PhantomData<B>),
```

Comment:
To me, `AnyOf` implies "zero or more of" instead of "one or more of". I went with `AtLeastOneOf` because I think it's clearer to explicitly spell out what the condition is. What about `SomeOf` as a compromise?

Reply:
It is in line with the semantics of [`Iterator::any`] though.

It can also be formulated as a question (since any is mostly used in questions rather than positive sentences):

```
ArchetypeStatement::<BundleA>::any_of()
```

can be interpreted as “Does the archetype have any component of `BundleA`?”. If the answer is “yes”, I would expect that the archetype has at least one component in common with the bundle. Otherwise, if the answer is “no”, I would expect that the archetype has zero components in common with the bundle.

Also, defining “any of” as “zero or more of” for countable/uncountable objects would not convey any information, since it would be true in any case.

[`Iterator::any`]: https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.any

Reply:
Alright, I'm convinced. `AnyOf` it is :). Fixed in eec04abc15e5e328c25c000bca42927876ec0d4d